prov/rxm/src/rxm_domain.c

```diff
@@ -372,9 +372,10 @@ static int rxm_mr_add_map_entry(struct util_domain *domain,
 	ofi_genlock_lock(&domain->lock);
 	ret = ofi_mr_map_insert(&domain->mr_map, msg_attr, &temp_key, rxm_mr, flags);
 	if (OFI_UNLIKELY(ret)) {
-		FI_WARN(&rxm_prov, FI_LOG_DOMAIN,
-			"MR map insert for atomic verification failed %d\n",
-			ret);
+		if (ret != -FI_ENOKEY)
+			FI_WARN(&rxm_prov, FI_LOG_DOMAIN,
+				"MR map insert for atomic verification failed %d\n",
+				ret);
```

Comment:
In the case of FI_ENOKEY, the key in the mr_map is still associated with an old rxm_mr, which apparently is still valid. We are now in a situation that a key is associated with multiple rxm_mr's, but only one is present in the mr_map.  

In `rxm_handle_atomic_req`,  `rxm_mr_get_map_entry` is called to retrieve the rxm_mr associated with the key. In this case, the old rxm_mr is going to be referenced. That may or may not work as expected. If one close the old rxm_mr. The key will be removed from the mr_map. The atomic req processing would fail even though it should.

I don't have a good solution for this. It's basically a behavior confliction of cached MR at verbs level and non-cached behavior of the mr_map at rxm level.

Reply:
would it make sense to do a lookup before inserting the rxm_mr and only insert if lookup fails?

Reply:
That doesn't make any difference. The mr_map is mainly used to find rxm_mr with a key. There can be only one association in mr_map, but we have multiple rxm_mr's.